gossipd/routing.c

```diff
@@ -523,68 +532,63 @@ bool handle_channel_announcement(
 		return false;
 	}
 
+	tag = type_to_string(tmpctx, struct short_channel_id,
+			     &short_channel_id);
+
 	/* BOLT #7:
 	 *
 	 * The receiving node MUST ignore the message if the specified
 	 * `chain_hash` is unknown to the receiver.
 	 */
 	if (!structeq(&chain_hash, &rstate->chain_hash)) {
-		status_trace("Received channel_announcement for unknown chain"
-			     " %s",
-			     type_to_string(tmpctx, struct bitcoin_blkid,
-					    &chain_hash));
+		status_trace(
+		    "Received channel_announcement %s for unknown chain %s",
+		    tag,
+		    type_to_string(tmpctx, struct bitcoin_blkid, &chain_hash));
 		tal_free(tmpctx);
 		return false;
 	}
 
 	// FIXME: Check features!
 	//FIXME(cdecker) Check chain topology for the anchor TX
 
-
 	local = pubkey_eq(&node_id_1, &rstate->local_id) ||
 		pubkey_eq(&node_id_2, &rstate->local_id);
+
 	sigfail = !check_channel_announcement(
 	    &node_id_1, &node_id_2, &bitcoin_key_1, &bitcoin_key_2,
 	    &node_signature_1, &node_signature_2, &bitcoin_signature_1,
 	    &bitcoin_signature_2, serialized);
 
 	status_trace("Received channel_announcement for channel %s, local=%d, sigfail=%d",
-		     type_to_string(trc, struct short_channel_id,
-				    &short_channel_id), local, sigfail);
+		     tag, local, sigfail);
 
-
-	if (sigfail && !local) {
+	if (sigfail) {
 		status_trace(
-		    "Signature verification of non-local channel announcement failed");
+			"Signature verification of channel_announcement for %s failed", tag);
 		tal_free(tmpctx);
 		return false;
 	}
 
-	/* Is this a new connection? */
+	/* Is this a new connection? It is if we don't know the
+	 * channel yet, or do not have a matching announcement in the
+	 * case of side-loaded channels*/
 	c0 = get_connection_by_scid(rstate, &short_channel_id, 0);
 	c1 = get_connection_by_scid(rstate, &short_channel_id, 1);
 	forward = !c0 || !c1 || !c0->channel_announcement || !c1->channel_announcement;
 
 	add_channel_direction(rstate, &node_id_1, &node_id_2, &short_channel_id,
-			      sigfail ? NULL : serialized);
+			      serialized);
 	add_channel_direction(rstate, &node_id_2, &node_id_1, &short_channel_id,
-			      sigfail ? NULL : serialized);
+			      serialized);
 
-	if (!forward || sigfail) {
-		status_trace("Not forwarding channel_announcement, forward=%d, sigfail=%d", forward, sigfail);
-		tal_free(tmpctx);
-		/* This will not be forwarded so we do not want to
-		 * announce the node either, others might drop it. */
-		return false;
+	if (forward) {
+		assert(!queue_broadcast(rstate->broadcasts, WIRE_CHANNEL_ANNOUNCEMENT,
+					(u8*)tag, serialized));
 	}
 
-	u8 *tag = tal_arr(tmpctx, u8, 0);
-	towire_short_channel_id(&tag, &short_channel_id);
-	assert(!queue_broadcast(rstate->broadcasts, WIRE_CHANNEL_ANNOUNCEMENT,
-				tag, serialized));
-
 	tal_free(tmpctx);
-	return local;
+	return local && forward;
```

Comment:
Do we still care that it's local?  The caller knows if it's local; I think this can be removed, along with the knowledge of our own id from this file...

Reply:
Actually the caller does not, the only entry point is `handle_gossip_msg` which handles all of the gossip related messages, only after decoding the message do we know whether it's local or not. `handle_local_add_channel` is what you're probably referring to, but that doesn't call `handle_channel_announcement`, and is not called with the fully signed message on the `funding_depth` callback ;-)

Reply:
Right.